types/react/index.d.ts

```diff
@@ -359,55 +359,156 @@ declare namespace React {
     // Component Specs and Lifecycle
     // ----------------------------------------------------------------------
 
-    interface ComponentLifecycle<P, S> {
+    // This should actually be something like `Lifecycle<P, S> | DeprecatedLifecycle<P, S>`,
+    // as React will _not_ call the deprecated lifecycle methods if any of the new lifecycle
+    // methods are present.
+    interface ComponentLifecycle<P, S> extends NewLifecycle<P, S>, DeprecatedLifecycle<P, S> {
+        /**
+         * Called immediately after a compoment is mounted. Setting state here will trigger re-rendering.
+         */
+        componentDidMount?(): void;
+        /**
+         * Called to determine whether the change in props and state should trigger a re-render.
+         *
+         * `Component` always returns true.
+         * `PureComponent` implements a shallow comparison on props and state and returns true if any
+         * props or states have changed.
+         *
+         * If false is returned, `Component#render`, `componentWillUpdate`
+         * and `componentDidUpdate` will not be called.
+         */
+        shouldComponentUpdate?(nextProps: Readonly<P>, nextState: Readonly<S>, nextContext: any): boolean;
+        /**
+         * Called immediately before a component is destroyed. Perform any necessary cleanup in this method, such as
+         * cancelled network requests, or cleaning up any DOM elements created in `componentDidMount`.
+         */
+        componentWillUnmount?(): void;
+        /**
+         * Catches exceptions generated in descendant components. Unhandled exceptions will cause
+         * the entire component tree to unmount.
+         */
+        componentDidCatch?(error: Error, errorInfo: ErrorInfo): void;
+    }
+
+    // Unfortunately, we have no way of declaring that the component constructor must implement this
+    interface StaticLifecycle<P, S> {
+        getDerivedStateFromProps?: GetDerivedStateFromProps<P, S>;
+    }
+
+    type GetDerivedStateFromProps<P, S> =
+        /**
+         * Returns an update to a component's state based on its new props and old state.
+         *
+         * Note: its presence prevents any of the deprecated lifecycle methods from being invoked
+         */
+        <K extends keyof S>(nextProps: Readonly<P>, prevState: Readonly<S>) => Pick<S, K> | S | null;
+
+    // This should be "infer SS" but can't use it yet
+    interface NewLifecycle<P, S, SS extends object = any> {
```

Comment:
This is my attempt to enforce that the type of the 3rd argument of `componentDidUpdate` is the same as the return value of `getSnapshotBeforeUpdate` minus `null`, but I'm not sure this will be useful in practice without adding a 3rd generic argument to `React.Component` (breaking change, would also require a 2nd argument on `ComponentType`) or falling back to `any`.

It would be ideal if the compiler would inherit the parameter and return types when declaring override functions...

Reply:
I believe that the snapshot returned can be anything, it doesn't have to be a proper object.

I can, say, return the scroll position of the window prior to a commit phase:

```tsx
class PreserveScroll extends React.Component {
  getSnapshotBeforeUpdate(): number {
    return document.documentElement.scrollHeight - window.scrollY
  }
  componentDidUpdate(_: any, _: any, scrollDifference: number) {
    window.scrollTo(document.documentElement.scrollHeight - scrollDifference)
  }
}
```